R/validate-target-data-utils.R

```diff
@@ -0,0 +1,183 @@
+# Target Data Validation Utilities
+
+## OBSERVABLE UNIT VALIDATION ----
+
+#' Validate observable_unit contains only valid columns
+#'
+#' Checks that the observable_unit array only contains:
+#' - Task ID columns (from tasks.json)
+#' - date_col value
+#'
+#' @param config_json Parsed target-data.json as list
+#' @param schema JSON schema for target-data
+#' @param task_id_names Character vector of task ID names
+#' @param level Character, one of "global", "time-series", or "oracle-output"
+#'
+#' @return NULL if validation passes, or data.frame with error row if validation fails
+#' @noRd
+validate_observable_unit <- function(
+  config_json,
+  schema,
+  task_id_names,
+  level = c("global", "time-series", "oracle-output")
+) {
+  level <- rlang::arg_match(level)
+  # Extract observable_unit based on level
+  if (level == "global") {
+    observable_unit <- config_json[["observable_unit"]]
+    path <- "observable_unit"
+  } else {
+    observable_unit <- config_json[[level]][["observable_unit"]]
+    path <- glue::glue("{level}/observable_unit")
+  }
+
+  # If no observable_unit defined, skip validation
+  if (is.null(observable_unit) || length(observable_unit) == 0) {
+    return(NULL)
+  }
+
+  # Get date_col from target data config
+  date_col <- get_date_col(config_json)
+
+  # Build allowed columns
+  allowed_cols <- unique(c(task_id_names, date_col))
+
+  # Check for invalid columns
+  invalid_cols <- setdiff(observable_unit, allowed_cols)
+
+  if (length(invalid_cols) > 0) {
+    # Format level for error message
+    level_label <- if (level == "global") "" else paste0(level, " ")
+
+    error_row <- data.frame(
+      instancePath = get_error_path(schema, path, "instance"),
+      schemaPath = get_error_path(schema, path, "schema"),
+      keyword = paste0(level_label, "observable_unit values"),
+      message = glue::glue(
+        "{level_label}observable_unit column(s) ",
+        "'{glue::glue_collapse(invalid_cols, \", \", last = \" & \")}' ",
+        "not valid. Must be task ID column(s) or date_col."
+      ),
+      schema = "",
+      data = glue::glue(
+        "observable_unit values: {glue::glue_collapse(observable_unit, ', ')}; ",
+        "allowed values: {glue::glue_collapse(allowed_cols, ', ')}"
+      ),
+      stringsAsFactors = FALSE
+    )
+    return(error_row)
+  }
+
+  NULL
+}
+
+
+## TIME-SERIES SPECIFIC VALIDATIONS ----
+
+#' Validate non_task_id_schema doesn't contain invalid columns
+#'
+#' Checks that non_task_id_schema doesn't contain:
+#' - Task ID columns
+#' - Reserved columns: as_of, output_type, output_type_id
```

Comment:
What about the date column (if it's not in the task ID columns)? Is it correct to add it to the non_task_id ?

Reply:
Good catched! Have updated the functionality to treat `date_col` as a reserved column and therefore flag it if it's included in the `non_task_id_schema` property here c40fe12a0ee7f2c0d0ced1abc36f838722144d1c